src/gsym/parser.rs

```diff
@@ -0,0 +1,339 @@
+        assert!(self.str_tab[self.str_tab.len() - 1] == 0);
+        assert!(off < self.str_tab.len());
+        unsafe {
+            CStr::from_ptr((&self.str_tab[off..]).as_ptr() as *const i8)
+                .to_str()
+                .unwrap()
+        }
+    }
+}
+
+/// Find the index of an entry in the address table most likely
+/// containing the given address.
+///
+/// The callers should check the respective `AddressInfo` to make sure
+/// it is what they request for.
+pub fn find_address(ctx: &GsymContext, addr: u64) -> usize {
+    let mut left = 0;
+    let mut right = ctx.num_addresses();
+
+    if right == 0 {
+        return 0;
+    }
+    if addr < ctx.addr_at(0) {
+        return 0;
+    }
+
+    while (left + 1) < right {
+        let v = (left + right) / 2;
+        let cur_addr = ctx.addr_at(v);
+
+        if addr == cur_addr {
+            return v;
+        }
+        if addr < cur_addr {
+            right = v;
+        } else {
+            left = v;
+        }
+    }
+    left
+}
+
+/// Parse AddressData.
+///
+/// AddressDatas are items following AndressInfo.
+/// [`GsymContext::addr_info()`] returns the raw data of AddressDatas as a
+/// slice at [`AddressInfo::data`].
+///
+/// # Arguments
+///
+/// * `data` - is the slice from AddressInfo::data.
+///
+/// Returns a vector of [`AddressData`].
+pub fn parse_address_data(data: &[u8]) -> Vec<AddressData> {
+    let mut data_objs = vec![];
+
+    let mut off = 0;
+    while off < data.len() {
+        let typ = decode_uword(&data[off..]);
+        off += 4;
+        let length = decode_uword(&data[off..]);
+        off += 4;
+        let d = &data[off..(off + length as usize)];
+        data_objs.push(AddressData {
+            typ,
+            length,
+            data: d,
+        });
+        off += length as usize;
+
+        #[allow(non_upper_case_globals)]
+        match typ {
+            InfoTypeEndOfList => {
+                break;
+            }
+            InfoTypeLineTableInfo | InfoTypeInlineInfo => {}
+            _ => {
+                eprintln!("unknown info type");
+            }
+        }
+    }
+
+    data_objs
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::env;
+    use std::fs::File;
+    use std::io::Read;
+    use std::path::Path;
+
+    #[test]
+    fn test_parse_context() {
+        let args: Vec<String> = env::args().collect();
+        let bin_name = &args[0];
+        let test_gsym = Path::new(bin_name)
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .join("data")
+            .join("test.gsym");
+        let mut gsym_fo = File::open(test_gsym).unwrap();
+        let mut data = vec![];
+
+        gsym_fo.read_to_end(&mut data).unwrap();
+        let ctx = parse_header(&data).unwrap();
+
+        assert_eq!(ctx.addr_at(1), 0x57280);
```

Comment:
I assume all those addresses are specific to the blob you checked-in? How was it generated and how can it be modified down the line without breaking all those tests?

Reply:
I run gsymutil on a test case runner generated by cargo for blazesym to create this blob.  These tests are very specific to this blob.  For example, check if a symbol is located at a specific address/location.  Without external input, these data should be hard coded to check their correctness.

However, I will comment on the test case to let people know how we get the blob.